benchmark/utils/ArgParse.swift

```diff
@@ -12,67 +12,200 @@
+        ? "'\(value)' is not a valid '\(type)'"
+        : "'\(value)' is not a valid '\(type)' for '\(argument!)'"
+    case let .unsupportedArgument(argument):
+      return "unsupported argument '\(argument)'"
+    }
+  }
+}
+
+/// Returns the argument value converted to the type T using the parser.
+/// If the parser cannot create the value of specified type, throw an invalid
+/// type argument error.
+func checked<T>(
+  _ parse: (String) throws -> T?,
+  _ value: String,
+  argument: String? = nil
+) throws -> T {
+  if let t = try parse(value)  { return t }
+  var type = "\(T.self)"
+  if type.starts(with: "Optional<") {
+      let s = type.index(after: type.index(of:"<")!)
+      let e = type.index(before: type.endIndex) // ">"
+      type = String(type[s..<e]) // strip Optional< >
   }
+  throw ArgumentError.invalidType(
+    value: value, type: type, argument: argument)
 }
 
-/// Using CommandLine.arguments, returns an Arguments struct describing
-/// the arguments to this program. If we fail to parse arguments, we
-/// return nil.
-///
-/// We assume that optional switch args are of the form:
-///
-/// --opt-name[=opt-value]
-/// -opt-name[=opt-value]
-///
-/// with opt-name and opt-value not containing any '=' signs. Any
-/// other option passed in is assumed to be a positional argument.
-public func parseArgs(_ validOptions: [String]? = nil)
-  -> Arguments? {
-  let progName = CommandLine.arguments[0]
-  var positionalArgs = [String]()
-  var optionalArgsMap = [String : String]()
-
-  // For each argument we are passed...
-  var passThroughArgs = false
-  for arg in CommandLine.arguments[1..<CommandLine.arguments.count] {
-    // If the argument doesn't match the optional argument pattern. Add
-    // it to the positional argument list and continue...
-    if passThroughArgs || !arg.starts(with: "-") {
-      positionalArgs.append(arg)
-      continue
+/// Parser that converts the program's command line arguments to typed values
+/// according to the parser's configuration, storing them in the provided
+/// instance of a value-holding type.
+class ArgumentParser<U> {
+    private var result: U
+    private var validOptions: [String] {
+      return arguments.compactMap { $0.name }
     }
-    if arg == "--" {
-      passThroughArgs = true
-      continue
+    private var arguments: [Argument] = []
+    private let programName: String = {
+      // Strip full path from the program name.
+      let r = CommandLine.arguments[0].reversed()
+      let ss = r[r.startIndex ..< (r.index(of:"/") ?? r.endIndex)]
+      return String(ss.reversed())
+    }()
+    private var positionalArgs = [String]()
+    private var optionalArgsMap = [String : String]()
+
+    // Argument holds the name of the command line parameter and the value
+    // processing closure used to convert it into given type and storing it
+    // in the parsing result.
+    struct Argument {
+      let name: String?
+      let help: String?
+      let apply: () throws -> ()
```

Comment:
The purpose of that closure (introduced in e5cbfcc) is to type erase the generic `T` from parsers, so that they can be stored in `[]`. The parsing happens as side effect of its execution. It being useful for printing help is just a pleasant accident.

Reply:
It isn't exactly obvious, so I'll add a better documentation for all of this.